### PR TITLE
First pass of refactoring filter_structures_plane

### PR DESCRIPTION
**Pull request by momchil-flex**
While further investigating inefficiencies in the case of a large number of structures, I realized that there's an inconsistency in our `Simulation.intersecting_media` method. The method promises to return intersections of the `test_object`, let's say a plane (if volume, each of the surfaces is recursively used) with a set of structures. However, what it actually does is return intersections of the **entire** plane in which the test object lies. That is to say, regardless of the in-plane size and position of my `test_object`, I'll always get the same set of intersections, many of which may actually not touch it at all.

This was an issue in a simulation with a large number of structures, and a `PlaneWave` source with a small and finite size that was actually not touching any of the structures, but was in the same plane as some. Granted, this is a strange setup (plane waves should usually have infinite size), but it is not something we are prohibiting. Even more importantly however, this can definitely happen with projection monitors, for which we are doing the same validation. In fact I discovered that one of our **tests** was wrong. It was computing the intersecting media between a volume projection monitor and a structure, and returning one intersection even though there are actually none, if the finite size of the two is properly taken into account.

As a proposed solution, I have refactored `_filter_structures_plane` to not take `x, y, z`, but instead a planar `Box`. This could also be useful in the future if we want to efficiently plot within some finite bounds smaller than the simulation domain. With the current methods, this will involve computing the intersections in the entire plane (could be slow) and only then setting the bounds. Instead, it would be more efficient to quickly exclude surfaces that are not touching the finite plane. Note that this would involve changing the `_get_structures_plane` method too.

This is work in progress. There are a few things that still need some thought:
- Should we refactor all intersections type methods to take a 2D box instead of `(x, y, z)`?
- In the updated implementation of `_filter_structures_plane`, it is still possible to get an intersection that does not actually exist. This is because I only exclude structures whose **bounding boxes** do not intersect the input box. However, you could imagine a PolySlab whose bounding box **does** intersect the box, but the polygon itself has no actual intersection with the box. I don't know if we can/want to handle this better.

### CODE REVIEW

**Comment by tylerflex** [comment]
> Should we refactor all intersections type methods to take a 2D box instead of (x, y, z)?

I feel like since these are public facing, we should leave them as is.

However, it could potentially be useful to provide alternatives that accept a box and do the heavy lifting. We could then wrap these methods with the x,y,z accepting methods and just construct boxes with size (0,0,inf) for example.

> In the updated implementation of _filter_structures_plane, it is still possible to get an intersection that does not actually exist. This is because I only exclude structures whose bounding boxes do not intersect the input box. However, you could imagine a PolySlab whose bounding box does intersect the box, but the polygon itself has no actual intersection with the box. I don't know if we can/want to handle this better.

I feel like this edge case is fine. The worst thing that can happen here is that it slows down the method slightly?

File: tidy3d/components/simulation.py [hunk 1]
```diff
@@ -1117,7 +1111,13 @@ def plot_structures_eps(  # pylint: disable=too-many-arguments,too-many-locals
             return ax
 
         if alpha < 1 and not isinstance(self.medium, CustomMedium):
-            medium_shapes = self._filter_structures_plane(structures=structures, x=x, y=y, z=z)
+            center = [0] * 3
+            size = [inf] * 3
+            axis, position = Box.parse_xyz_kwargs(x=x, y=y, z=z)
+            center[axis] = position
+            size[axis] = 0
```
Comment (tylerflex) [inline tidy3d/components/simulation.py:1118]: cleaner way to do this:
```py
center = Box.unpop_axis(position, (0, 0), axis=axis)
size = Box.unpop_axis(0, (inf, inf), axis=axis)
```

File: tidy3d/components/simulation.py [hunk 2]
```diff
@@ -1810,11 +1810,15 @@ def _filter_structures_plane(  # pylint:disable=too-many-locals
         for structure in structures:
 
             # dont bother with geometries that dont intersect plane
-            if not structure.geometry.intersects_plane(x=x, y=y, z=z):
+            if not structure.geometry.intersects(plane):
```
Comment (tylerflex) [inline tidy3d/components/simulation.py:1813]: I think this reads like a `Geometry.intersections(plane)` method yea.

So right now we have
```py
Geometry.intersections(x,y,z) -> List[Shapely]
Geometry.intersects_plane(x,y,z) -> bool
```
maybe we should in fact introduce a couple analogues that accept Box as arguments.
Not sure what to name them, but if we could start over, I'd say
```py
Geometry.intersections(x,y,z) -> List[Shapely]
Geometry.intersections_plane(plane: Box) -> List[Shapely]
Geometry.intersects(x,y,z) -> bool
Geometry.intersects_plane(plane: Box) -> bool
```
although this breaks backwards compatibility.
Maybe for the plane accepting ones, we could do `Geometry.interesects_plane_box` and `Geometry.intersections_box`. For example?

File: tidy3d/components/simulation.py [hunk 3]
```diff
@@ -1784,7 +1784,7 @@ def _get_structures_plane(
 
     @staticmethod
     def _filter_structures_plane(  # pylint:disable=too-many-locals
-        structures: List[Structure], x: float = None, y: float = None, z: float = None
+        structures: List[Structure], plane: Box
```
Comment (tylerflex) [inline tidy3d/components/simulation.py:1787]: I think we need to ensure that the plane is in fact 2D and error if not, especially in the public facing methods.

**Review by tylerflex — CHANGES_REQUESTED** [review]
Looks good, I feel like we should merge this in ASAP if it helps with the performance issues. What I dont see off the top of my head is:
1. Does changing _filter_structures_plane break anything else? eg in the backend?
2. Are the existing intersections methods slowing down anything else when used internally? If so, we should think about changing their APIs or adding `box` accepting ones. Maybe out of backwards compatibility concerns we should hold off on this until 2.0 and just merge the existing changes.

**Comment by shashwat-sh** [comment]
It looks like the error is related to source amp / flux computation in the backend, when we try get an epsilon value on the source plane. It should be possible to reproduce this in a frontend test I would think, where a plane intersects a custom medium.  I'll see if I can add a frontend test to this effect.

**Comment by momchil-flex** [comment]
Actually could you check that it is indeed this commit that broke the test and not my previous inside/intersection updates that are merged in?

**Comment by shashwat-sh** [comment]
> Actually could you check that it is indeed this commit that broke the test and not my previous inside/intersection updates that are merged in?

You're right, [this](732c00a45ad3571691c44616accad82414df7913) commit already breaks the custom medium backend test.

**Comment by momchil-flex** [comment]
Alright, I'm going to sort this out today.

File: tidy3d/components/simulation.py (hunk 1 above)
Comment (momchil-flex) [reply tidy3d/components/simulation.py:1118]: noice

**Comment by momchil-flex** [comment]
I implemented all the changes and fixed the backend tests with custom medium (which were broken by my previous intersections PR).

The `test_sim_monitor_homogeneous` turned out to be all sorts of wrong, which caught me off guard and took me a while to figure out what's up. Basically, it was producing a SetupError not because of the monitors, but because of the PlaneWave source! Weirdly, the PlaneWave source's position was moved for the final test to make it work... Anyway, I replaced the source with a PointDipole, and adjusted the boxes sizes a bit for the test to actually work as intended.

> I feel like this edge case is fine. The worst thing that can happen here is that it slows down the method slightly?

Actually I fixed this edge case by intersecting all of the found intersection shapes with the input `plane` at the end.

The only thing is naming, yeah. In my current implementation we have:

```
Geometry.intersections(x, y, z) -> List[Shapely]
Geometry.intersects(other: Geometry) -> bool # note: existing already
Geometry.intersects_plane(x, y, z) -> bool
Geometry.intersections_plane(plane: Box) -> List[Shapely] # new
```

Obviously not ideal - thoughts?

**Comment by tylerflex** [comment]
@momchil-flex see edited comment

**Comment by shashwat-sh** [comment]
> The test_sim_monitor_homogeneous turned out to be all sorts of wrong, which caught me off guard and took me a while to figure out what's up. Basically, it was producing a SetupError not because of the monitors, but because of the PlaneWave source! Weirdly, the PlaneWave source's position was moved for the final test to make it work... Anyway, I replaced the source with a PointDipole, and adjusted the boxes sizes a bit for the test to actually work as intended.

Entirely my fault - sorry about that. I probably added the final test separately from the rest, when taking `exclude_surfaces` into account, which might be why the plane wave thing wasn't caught.

**Comment by momchil-flex** [comment]
> I feel like ideally we denote `plane` to signify the method accepts one of `xyz` position inputs and then the "plane-less" methods accept regular boxes or other geometries. Maybe the following works?
> 
> ```python
> Geometry.intersects_plane(x, y, z) -> bool
> Geometry.intersections_plane(x, y, z) -> List[Shapely]
> 
> Geometry.intersects(other: Geometry) -> bool # note: existing already
> Geometry.intersections_box2d(plane: Box) -> List[Shapely] # new, validates that plane has exactly one size == 0
> Geometry.intersections(other: Geometry) -> List[Shapely] # down the line, if we want a more general method we can reserve just pure `Geometry.intersect*` to accept `Geometry`?
> Geometry.intersects_box2d(plane: Box) -> bool # down the line, for completeness
> ```
> 
> So I guess in summary
> 
> ```python
> Geometry.intesect*(other: Geometry)
> Geometry.intesect*_plane(x: float = None, y: float = None, z: float = None)
> Geometry.intesect*_box2d(plane: Box)
> ```

This makes sense conceptually, however:

- `Geometry.intersects_box2d(plane)` is not really necessary as it's handled by `Geometry.intersects(other)`.
- `Geometry.intersections(other)` is unclear for 3D `other`. I guess it could be intersections with the surfaces of `other`? But in any case yeah probably not something to do now
- I guess the main question is whether to rename `intersections` to `intersections_plane` even though it breaks compatibility. Unlikely that anyone is using this externally right now so I'd be fine with that...

**Comment by tylerflex** [comment]
> Geometry.intersects_box2d(plane) is not really necessary as it's handled by Geometry.intersects(other)

Yea this would just be for convenience / completeness and could just wrap `Geometry.intersects(other)`. Not saying we include it, necessarily, just writing out what could follow based on the naming proposal where the suffix of the method name corresponds to the type of input expected.

> Geometry.intersections(other) is unclear for 3D other. I guess it could be intersections with the surfaces of other? But in any case yeah probably not something to do now

Yea, again just seeing how the naming would work out, but not saying we do it now. In fact, let's not :D 

> I guess the main question is whether to rename intersections to intersections_plane even though it breaks compatibility. Unlikely that anyone is using this externally right now so I'd be fine with that...

Hm, I guess we could do the deprecation warning stuff and fix it for 2.0 to be safe?

**Comment by momchil-flex** [comment]
> > I guess the main question is whether to rename intersections to intersections_plane even though it breaks compatibility. Unlikely that anyone is using this externally right now so I'd be fine with that...
> 
> Hm, I guess we could do the deprecation warning stuff and fix it for 2.0 to be safe?

Alright, I renamed all `intersections` to `intersections_plane` but left an `intersections` method that issues a deprecation warning then calls `intersections_plane`. This way the warning will not be raised when we call `intersections_plane` directly internally.

I left this as a separate commit but generally this seems good to go to me now, will just rebase and add to changelog when I get the OK from you.

**Comment by tylerflex** [comment]
ok sounds good, but just to clarify, what will be the final list of intersect_like method call signatures?

**Comment by momchil-flex** [comment]
Basically the logic will be like you described, but some will not be implemented

```
Geometry.intersects_plane(x, y, z) -> bool
Geometry.intersections_plane(x, y, z) -> List[Shapely]
Geometry.intersects(other: Geometry) -> bool # note: existing already
Geometry.intersections(other: Geometry) -> List[Shapely] # will not exist in 2.0 
Geometry.intersections_box2d(plane: Box) -> List[Shapely]
Geometry.intersects_box2d(plane: Box) -> bool # not implemented as it would just wrap `intersects`
```

**Comment by tylerflex** [comment]
k looks good to me, you may proceed!

**Review by shashwat-sh — COMMENTED** [review]
Looks good to me too.

File: CHANGELOG.md [hunk 4]
```diff
@@ -14,12 +14,14 @@ run time expected as a result.
...
 
 ### Changed
+- Issue a deprecation warning that `Geometry.intersections` will be renamed to `Geometry.intersections_plane` in 2.0.
 
 
 ### Fixed
 - Progressbars always set to 100% when webapi functions are finished.
 - Faster handling of `Geometry.intersects` and `Geometry.inside` by taking into account geometry bounds.
 - Numpy divide by zero warning in mode solver fixed by initializing jacobians as real instead of complex.
+- Bug in validators for 2D objects being in homogeneous media which were looking at the infinite plane in which the objects lie.
```
Comment (tylerflex) [inline CHANGELOG.md:24]: feel like you can make this a bit stronger, to mention that there's a big speedup for many structures?
